plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeIFEInputSpent.sol

```diff
@@ -0,0 +1,87 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "../PaymentExitDataModel.sol";
+import "../PaymentInFlightExitModelUtils.sol";
+import "../routers/PaymentInFlightExitRouterArgs.sol";
+import "../spendingConditions/IPaymentSpendingCondition.sol";
+import "../spendingConditions/PaymentSpendingConditionRegistry.sol";
+import "../../utils/ExitId.sol";
+import "../../../utils/UtxoPosLib.sol";
+import "../../../utils/Merkle.sol";
+import "../../../vaults/EthVault.sol";
+import "../../../framework/PlasmaFramework.sol";
+import "../../../transactions/PaymentTransactionModel.sol";
+
+library PaymentChallengeIFEInputSpent {
+    using UtxoPosLib for UtxoPosLib.UtxoPos;
+    using PaymentInFlightExitModelUtils for PaymentExitDataModel.InFlightExit;
+
+    // TODO: Use BondSize lib.
+    uint256 public constant PIGGYBACK_BOND = 31415926535 wei;
+
+    struct Controller {
+        PlasmaFramework framework;
+        PaymentSpendingConditionRegistry spendingConditionRegistry;
+        uint256 supportedTxType;
+        EthVault ethVault;
+    }
+
+    event InFlightExitInputBlocked(
+        address indexed challenger,
+        bytes32 txHash,
+        uint8 inputIndex
+    );
+
+    function run(
+        Controller memory self,
+        PaymentExitDataModel.InFlightExitMap storage inFlightExitMap,
+        PaymentInFlightExitRouterArgs.ChallengeInputSpentArgs memory args
+    )
+        public
+    {
+        uint192 exitId = ExitId.getInFlightExitId(args.inFlightTx);
+        PaymentExitDataModel.InFlightExit storage ife = inFlightExitMap.exits[exitId];
+
+        require(ife.exitStartTimestamp != 0, "In-flight exit doesn't exist");
+
+        require(ife.isInputPiggybacked(args.inFlightTxInputIndex), "The indexed input has not been piggybacked");
+
+        require(
+            keccak256(args.inFlightTx) != keccak256(args.spendingTx),
+            "The spending transaction is the same as the in-flight transaction"
+        );
+
+        PaymentTransactionModel.Transaction memory spendingTx = PaymentTransactionModel.decode(args.spendingTx);
```

Comment:
Hmm...I think we cannot really assume the spending tx would be a Payment tx. Imagine a ERC721 tx is your input tx.

However, I think this check + the following (checking input pointing to the right output) should be done in SpendingCondition.

Reply:
So should the spendingTx type be passed in as a parameter?

BTW is your argument true for this line as well?
```
IPaymentSpendingCondition condition = self.spendingConditionRegistry.spendingConditions(
```
If we don't know it's a Payment tx, then we can't use a PaymentSpendingCondition?

Reply:
hmmm, yeah I guess we would need the type to be passed in.
As for second question, that is why this issue: https://github.com/omisego/plasma-contracts/issues/214 and that is what I am working on right now.

Reply:
Ok.
However, we still need to decode the spendingTx to be able to check that the specified input really is the same as the piggybacked input. Should we use WireTransaction for this?

Reply:
This is what I am thinking:
1. In ExitGame, build the `outputId` using the `inputTx + outputIndex(of input tx) + utxoPos of inputTx` (Spending Condition input). Verify the outputId is same as what we saved in exit data of IFE (we have `outputId` in WithdrawData). This makes sure we are using the same output of the IFE input.
1. Inside SpendingCondition, it would verify whether the spending tx is pointing to the output as expected. So we would make sure the spending tx is pointing to the output from the data given (`inputTx, outputIndex, txPos`)
1. So instead of directly checking spendingTx and IFE tx have same pointer, we verify both are pointing to same output. Though one is done inside ExitGame and the other done in SpendingCondition.

Reply:
Sorry, not really sure what you're getting at 🤔 
Which `inputTx` you're referring to here?

- In PaymentStartInFlightExit the `outputId` of the IFE tx input is built and put into WithdrawData
- In PaymentChallengeIFEInputSpent, the inFlightTx is passed in so `outputId` of the IFE tx input can be retrieved using the `exitId`.
- Are you saying that the the caller has to pass in the `outputId` of the spending tx input, so that it can be passed into SpendingCondition.verify?

Reply:
> Which inputTx you're referring to here?

Ah...I was in the mind of using new generalized spending condition interface. Wasn't realized it's never showed up in current code here. That is the input tx of ife tx and spending tx.

> Are you saying that the the caller has to pass in the outputId of the spending tx input, so that it can be passed into SpendingCondition.verify?

Ah, again, I was in the mind set of #214.
So not really the `outputId` directly, but the original data for `outputId` (txBytes, outputIndex, utxoPos).
(can be build from spending condition args: bytes inputTx, uint16 outputIndex,  uint256 inputTxPos)

---
hmm, let's me rephrase a bit. So the main problem here is that both input tx (of ife) and spending tx (of input tx) can be not `PaymentTx`. However, we want to be able to check whether the spending tx is pointing to the same input as the in-flight exit tx is pointing to.

instead of
```
require(ifeInput == spendingInput, "Spent input is not the same as piggybacked input")
```

I am saying we try to the following in exit game
```
// outputId is from Withdraw data, has same value of "ifeInput" as above 
// building the outputId itself does not require the need to decode transaction, so it is fine to do this in ExitGame
require(outputId == buildedOutputIdFromSpendingConditionArgs, "....");
```
and then the spending condition would do the sth like
```
// buildedOutputIdentifierFromSpendingConditionArgs can be outputId or utxoPos, depending on spendingTx type
require(buildedOutputIdentifierFromSpendingConditionArgs == spendingTx.inputs[index], "....")
```

SInce `buildedOutputIdFromSpendingConditionArgs` and `buildedOutputIdentifierFromSpendingConditionArgs` are both from same args of spending condition, this makes sure they are indicating same thing. So with the two requirements above, it would represent the same thing as `outputId == spendingTx.inputs[index]`.

Reply:
Ok, thanks for clarifying.
I'll hold off on this until SpendingCondition is done...